doc/developers/contributing.rst

```diff
@@ -251,11 +251,21 @@ If any of the above seems like magic to you, then look up the `Git documentation
 web, or ask a friend or another contributor for help.
 
 If some conflicts arise between your branch and the ``master`` branch, you need
-to merge ``master``. The command will be::
+to merge ``master``. For that, you first need to fetch the ``upstream``'s
+``master``, and then merge it into your branch::
 
-  $ git merge master
+  $ git fetch upstream
+  $ git merge upstream/master
 
-with ``master`` being synchronized with the ``upstream``.
+It is recommended to work on temporary branches rather than directly on
+``master``. Since these temporary branches are branched off ``master``, it is
+helpful to keep your local ``master`` up to date with the upstream's
+``master``. To do this, you can::
+
+  $ git fetch upstream
+  $ git rebase upstream/master
```

Comment:
well, I would usually do `git reset --hard upstream/master` here.

Reply:
+1 although it's not necessary to ever use update the local `master` branch if all the feature branches are branched from `upstream/master` as recommended in the previous comment.

Reply:
> well, I would usually do `git reset --hard upstream/master` here.

`reset --hard` is fine as long as the user really knows they're not in their feature branch. I rather not have a command here which could be copy pasted by accident and result in the user loosing their changes.